src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1081,6 +1082,74 @@ private void ApplyTextDocumentChange(DocumentId documentId, SourceText newText)
             }
         }
 
+        protected override void ApplyDocumentInfoChanged(DocumentId id, DocumentInfo updatedInfo)
+        {
+            var document = CurrentSolution.GetDocument(id);
+
+            FailIfDocumentInfoChangesNotSupported(document, updatedInfo);
+
+            if (document.Name != updatedInfo.Name)
+            {
+                GetProjectData(updatedInfo.Id.ProjectId, out var _, out var project);
+
+                var projectItemForDocument = project.FindItemByPath(document.FilePath, StringComparer.OrdinalIgnoreCase);
+
+                if (projectItemForDocument == null)
+                {
+                    // TODO: The document might actually belong to a Shared Project.
+                    return;
+                }
+
+                // Must save the document first for things like Breakpoints to be preserved.
+                projectItemForDocument.Save();
+
+                var uniqueName = projectItemForDocument.Collection.GetUniqueName(
+                    Path.GetFileNameWithoutExtension(updatedInfo.Name),
+                    Path.GetExtension(updatedInfo.Name));
+
+                // By setting this property, Visual Studio will perform the file rename, which 
+                // will cause the workspace's current solution to update and will fire the 
+                // necessary workspace changed events.
+                projectItemForDocument.Name = uniqueName;
+            }
+        }
+
+        /// <summary>
+        /// The VisualStudioWorkspace currently supports only a subset of DocumentInfo changes.
+        /// </summary>
+        private void FailIfDocumentInfoChangesNotSupported(CodeAnalysis.Document document, DocumentInfo updatedInfo)
```

Comment:
can we do a onetime reflection over the type (maybe lazily) to get properties and compare that way?

Reply:
I don't have a good answer to this. I'm pretty opposed to reflecting over DocumentAttributes but I don't have anything better.

If we do nothing now, and in the future we forget to update this method when DocumentInfo updates (which means Document/DocumentAttributes changes), then someone might make try to make a corresponding change to the workspace and nothing happens. It would be a bug.

Open to more suggestions.

Reply:
What about implemented a `HasChanges(Document)` method on `DocumentInfo` and moving this code to there? Or implementing `IEqualityfComparer<Document>`. At least then it's closer to where changes would happen, rather than comparison being in a completely different location. 

Other ideas, maybe: 

Attributes to mark properties that will be compared.
A static instance of a comparer type. 
A static list of property names that get compared. 

Anything to help changes in the future do the right thing, really... Right now it would be hard for someone to find that they need to update this method.

Reply:
So if you have an equality and With* methods, one thing you can do is say that if you have:

```
DocumentInfo a;
DocumentInfo b;

bool differsOnlyByFoo = a.Equals(b.WithFoo(a.Foo));
```

Not sure how well that works though since DocumentInfo holds some hard-to-define equality things.